docs/design-proposals/multiple-statefulsets.md

```diff
@@ -0,0 +1,417 @@
+
+The TiDB Operator will create extra Services together with multiple statefulsets:
+
+* For PD: no change
+* For TiKV: no change
+* For TiDB: N+1 Services should be created, N client services and 1 main client service
+
+### Upgrade:
+
+With multiple statefulsets supported, the user can modify the spec in the `values.yaml` then the Operator will range over all the statefulsets and select only one statefulset to upgrade at a time. So the basic progress is the same with older design.
+
+The user can modify all the statefulsets, but we only support upgrade one Pod at a time.
+
+For example, the user can change the images from v2.1.0(Figure 5) to v2.2.0(Figure 6) to upgrade the cluster from v2.1.0 to v2.2.0:
+
+Figure 5:
+
+``` yaml
+tikv:
+  replicas: 3
+  image: pingcap/tikv:v2.1.0
+  …
+
+tikvs:
+  - name: name-1
+    replicas: 2
+    image: pingcap/tikv:v2.1.0
+    …
+  - name: name-2
+    replicas: 2
+    image: pingcap/tikv:v2.1.0
+    …
+```
+
+Figure 6:
+
+``` yaml
+tikv:
+  replicas: 3
+  image: pingcap/tikv:v2.2.0
+  …
+
+tikvs:
+  - name: name-1
+    replicas: 2
+    image: pingcap/tikv:v2.2.0
+    …
+  - name: name-2
+    replicas: 2
+    image: pingcap/tikv:v2.2.0
+    …
+```
+
+The operator does not guarantee the order of multiple statefulsets’ upgrade. If the user care about the order indeed. Just: modify one statefulset, wait it done, and then modify the other one.
+
+### Scale In/Out:
+
+Like the Upgrade section, the user can sale in/out all the statefulsets at the same time, operator will only scale in/out one Pod at a time.
+The user can just modify one statefulset if they want to operate the special statefulset.
+
+For example, the user can change the replicas in name-1 from 3(Figure 7) to 5(Figure 8) to add more replicas in the name-1:
+
+Figure 7:
+
+``` yaml
+tikv:
+  replicas: 3
+  image: pingcap/tikv:v2.1.0
+  …
+
+tikvs:
+  - name: name-1
+    replicas: 3
+    image: pingcap/tikv:v2.1.0
+    …
+  - name: name-2
+    replicas: 2
+    image: pingcap/tikv:v2.1.0
+    …
+```
+
+Figure 8:
+
+``` yaml
+tikv:
+  replicas: 3
+  image: pingcap/tikv:v2.1.0
+  …
+
+tikvs:
+  - name: name-1
+    replicas: 5
+    image: pingcap/tikv:v2.1.0
+    …
+  - name: name-2
+    replicas: 2
+    image: pingcap/tikv:v2.1.0
+    …
+```
+
+The operator does not guarantee the order of multiple statefulsets’ scale in/out. If the user care the order indeed. Just: modify one statefulset, wait it done, and then modify the other one.
+
+### Failover:
+
+When a PD/TiKV/TiDB peer is failure, the failover module will add a new peer to the failed statefulset. Others are the same with older design.
+
+### Vertical scaling
+
+To do vertical scaling, the user must:
+
+- Add a new statefulset in the `tikvs` section, using more `requests` or `limits` and wait it done. Add a new `name-3`, change from figure 9 to figure 10;
+- Scale down the old statefulset(`name-1`) in the `tikvs` section to zero and wait it done. See figure 11;
```

Comment:
Do we need to wait for the new statefulset name-3 to start completely before scale down

Reply:
Yes.

Reply:
**The user** must ensure that.